assembly-syntax/src/ast/type.rs

```diff
@@ -52,6 +57,454 @@ impl From<EnumType> for TypeDecl {
+
+/// A syntax-level type expression (i.e. primitive type, reference to nominal type, etc.)
+#[derive(Debug, Clone, Eq, PartialEq, Hash)]
+pub enum TypeExpr {
+    /// A primitive integral type, e.g. `i1`, `u16`
+    Primitive(Span<Type>),
+    /// A pointer type expression, e.g. `*u8`
+    Ptr(PointerType),
+    /// An array type expression, e.g. `[u8; 32]`
+    Array(ArrayType),
+    /// A struct type expression, e.g. `struct { a: u32 }`
+    Struct(StructType),
+    /// A reference to a type aliased by name, e.g. `Foo`
+    Ref(Ident),
+}
+
+impl From<Type> for TypeExpr {
+    fn from(ty: Type) -> Self {
+        match ty {
+            Type::Array(t) => Self::Array(ArrayType::new(t.element_type().clone().into(), t.len())),
+            Type::Struct(t) => {
+                Self::Struct(StructType::new(t.fields().iter().enumerate().map(|(i, ft)| {
+                    let name = Ident::new(format!("field{i}")).unwrap();
+                    StructField {
+                        span: SourceSpan::UNKNOWN,
+                        name,
+                        ty: ft.ty.clone().into(),
+                    }
+                })))
+            },
+            Type::Ptr(t) => Self::Ptr((*t).clone().into()),
+            Type::Function(_) => {
+                Self::Ptr(PointerType::new(TypeExpr::Primitive(Span::unknown(Type::Felt))))
+            },
+            Type::List(t) => Self::Ptr(
+                PointerType::new((*t).clone().into()).with_address_space(AddressSpace::Byte),
+            ),
+            Type::I128 | Type::U128 => Self::Array(ArrayType::new(Type::U32.into(), 4)),
+            Type::I64 | Type::U64 => Self::Array(ArrayType::new(Type::U32.into(), 2)),
+            Type::Unknown | Type::Never | Type::F64 => panic!("unrepresentable type value: {ty}"),
+            ty => Self::Primitive(Span::unknown(ty)),
+        }
+    }
+}
+
+impl Spanned for TypeExpr {
+    fn span(&self) -> SourceSpan {
+        match self {
+            Self::Primitive(spanned) => spanned.span(),
+            Self::Ptr(spanned) => spanned.span(),
+            Self::Array(spanned) => spanned.span(),
+            Self::Struct(spanned) => spanned.span(),
+            Self::Ref(spanned) => spanned.span(),
+        }
+    }
+}
+
+impl crate::prettier::PrettyPrint for TypeExpr {
+    fn render(&self) -> crate::prettier::Document {
+        use crate::prettier::*;
+
+        match self {
+            Self::Primitive(ty) => display(ty),
+            Self::Ptr(ty) => ty.render(),
+            Self::Array(ty) => ty.render(),
+            Self::Struct(ty) => ty.render(),
+            Self::Ref(ty) => display(ty),
+        }
+    }
+}
+
+// POINTER TYPE
+// ================================================================================================
+
+#[derive(Debug, Clone)]
+pub struct PointerType {
+    pub span: SourceSpan,
+    pub pointee: Box<TypeExpr>,
+    addrspace: Option<AddressSpace>,
+}
```

Comment:
Not for this PR, but would be good to add doc comments for this struct (as well as other structs in this module). One thing that is not immediately clear is what is the default address space (after looking further in the code, I think it is `Element`).

Also, nit: why make `addrspace` optional? If we assume that the default is `Element` - might as well just set it at construction time.

Reply:
It's optional syntax, but the actual address space is always known - this type is an AST structure, so it reflects the syntax used.

As for why we might allow the syntax to be optional - it makes for much more succinct type signatures when pointers are involved, e.g. `ptr<T>` instead of `ptr<T, addrspace(felt)>`